### PR TITLE
yubikeytoken: fix signature generation, add signature validation 

### PR DESCRIPTION
**Pull request by jh23453**
Please check the four commits:
- fix signature generation: thinko fixed
- add signature validation function: unused util the last patch is applied
- fix python backtrace / SQL error 
- add signature validation logic

### CODE REVIEW

**Comment by cornelinux** [comment]
Hi Jochen,
thanks a lot for the patch.
There is a Travis CI test failing.

I will not be able to fix this and merge this today. I think I can do this during the next days.

But also for future pull requests you should get familiar with testing the code before committing it and issuing the pull request.

You should be able to do it like this:

    cd [pi-srcdirectory]
    virtualenv venv
    source venv/bin/activate
    pip install -r requirements.txt
    py.test tests/

Then you need to wait for 10 or 20 minutes and you will see, if everything is fine ;-)
Kind regards
Cornelius

**Comment by cornelinux** [comment]
There will be a release next week, so I will not add the pull request into this release anymore.

**Comment by jh23453** [comment]
Cornelius Kölbel <notifications@github.com> writes:

> Hi Jochen,
> thanks a lot for the patch.
> There is a Travis CI test failing.
>
> I will not be able to fix this and merge this today. I think I can do this during the next days.

I think the following patch should fix the failing test:

diff --git a/tests/test_lib_tokens_yubikey.py b/tests/test_lib_tokens_yubikey.py
index 6d5916f..3687c23 100644
--- a/tests/test_lib_tokens_yubikey.py
+++ b/tests/test_lib_tokens_yubikey.py
@@ -153,7 +153,7 @@ class YubikeyTokenTestCase(MyTestCase):
                 "nonce": "blablafoo",
                 "timestamp": "time"}
         h = YubikeyTokenClass._api_signature(data, api_key)
-        self.assertEqual(h, "InGUM9l0cgfOzEAb0C1LgMktycU=")
+        self.assertEqual(h, "0KRJecfPNSrpZ79+xODbJl0HM8I=")
 
     def test_10_api_endpoint(self):
         fixed = "ebedeeefegeheiej"

> But also for future pull requests you should get familiar with testing the code before committing it and issuing the pull request.

Yes, that should help :-)
>
> You should be able to do it like this:
>
>     cd [pi-srcdirectory]
>     virtualenv venv
>     source venv/bin/activate
>     pip install -r requirements.txt
>     py.test tests/
>
> Then you need to wait for 10 or 20 minutes and you will see, if everything is fine ;-)

I needed to install a nice number of packages and got the tests running.
BTW, the README.rst has the following concerning the tests:

Run tests
=========

    nosetests -v --with-coverage --cover-package=privacyidea --cover-html

Jochen

-- 
The only problem with troubleshooting is that the trouble shoots back.

**Comment by jh23453** [comment]
Cornelius Kölbel <notifications@github.com> writes:

> There will be a release next week, so I will not add the pull request into this release anymore.

That's fine with me - I can run my patches locally until the next
release without problems.

Jochen

-- 
The only problem with troubleshooting is that the trouble shoots back.

**Comment by cornelinux** [comment]
you can either run nosetest or py.test.
Lately I prefer running py.test.

File: privacyidea/lib/tokens/yubikeytoken.py [hunk 1]
```diff
@@ -372,6 +372,8 @@ def api_endpoint(cls, request, g):
             data['status'] = "NO_SUCH_CLIENT"
             data['h'] = ""
         elif otp and id and nonce:
+            options = {"g": g,
```
Comment (cornelinux) [inline privacyidea/lib/tokens/yubikeytoken.py]: The options are not used anywhere!?

Comment (jh23453) [inline privacyidea/lib/tokens/yubikeytoken.py]: Cornelius Kölbel <notifications@github.com> writes:

>> @@ -372,6 +372,8 @@ def api_endpoint(cls, request, g):
>>              data['status'] = "NO_SUCH_CLIENT"
>>              data['h'] = ""
>>          elif otp and id and nonce:
>> +            options = {"g": g,
>
> The options are not used anywhere!?

Yes, that's what I figured out, but now got a message that parameter g
in that function is unused.  I'll try to figure out who is calling with
what parameters...

For now, please disregard the pull request - I'll prepare another one
after release.

Jochen

-- 
The only problem with troubleshooting is that the trouble shoots back.

**Comment by cornelinux** [comment]
Hi Jochen,

I merged the previous state for you pull request.
https://github.com/privacyidea/privacyidea/commit/47e64327a3c4dff09e56a101de6f5cbcfeb17477

So I just close this one.

Quantified code sometimes complains about unused parameters. That is correct.
But the base function api_endpoint has the "g" parameter, since some tokens might need this.
But the yubikey does not. So let quantified code complain. ;-)

Thanks a lot and kind regards
Cornelius

**Pull request closed without merging**